### PR TITLE
implementation of DeRec cryptography interface

### PR DESCRIPTION
**Pull request by rsinha**
This PR addresses the following:

- uses elliptic curve cryptography instead of RSA. Signing uses ECDSA and encryption uses ECIES, both with p256 curve.
- implements a sign-then-encrypt functionality, which implements a sign_then_encrypt function and a decrypt_then_verify function.
- implements a JNI interface for all cryptography primitives.

### CODE REVIEW

File: src/secret_sharing/utils.rs [hunk 1]
```diff
@@ -32,6 +32,7 @@ pub fn compute_sha256_hash(input: &[u8]) -> Vec<u8> {
     hasher.finalize().to_vec()
 }
 
+#[allow(dead_code)]
 pub fn prg_seed_expansion(rand: &[u8]) -> Option<[u8; 32]> {
     if rand.len() != 16 { //expecting only 128 bit keys
```
Comment (johnalanwoods) [inline src/secret_sharing/utils.rs:37]: Why do we expect only 128-bit keys?

Comment (rsinha) [reply src/secret_sharing/utils.rs:37]: I think we should eliminate this function altogether. It's being invoked by the ADSS construction, that we are not using anymore. And if we were to retain it, we could perhaps extract 256 bits directly from its invocation of the random oracle.

Comment (rsinha) [reply src/secret_sharing/utils.rs]: @johnalanwoods Added a commit to remove the dead code mentioned above.

**Review by lbaird — APPROVED** [review]

**Comment by johnalanwoods** [comment]
Approved

**Pull request merged**